service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/workspace/LaunchCloneControlledAzureStorageContainerResourceFlightStep.java

```diff
@@ -0,0 +1,101 @@
+      ControlledAzureStorageContainerResource sourceResource,
+      String subflightId,
+      UUID destinationResourceId,
+      @Nullable UUID destinationFolderId) {
+    this.sourceResource = sourceResource;
+    this.subflightId = subflightId;
+    this.destinationResourceId = destinationResourceId;
+    this.destinationFolderId = destinationFolderId;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    validateRequiredEntries(
+        context.getInputParameters(),
+        JobMapKeys.AUTH_USER_INFO.getKeyName(),
+        ControlledResourceKeys.DESTINATION_WORKSPACE_ID);
+
+    var userRequest =
+        context
+            .getInputParameters()
+            .get(JobMapKeys.AUTH_USER_INFO.getKeyName(), AuthenticatedUserRequest.class);
+    var destinationWorkspaceId =
+        context
+            .getInputParameters()
+            .get(ControlledResourceKeys.DESTINATION_WORKSPACE_ID, UUID.class);
+
+    // build input parameter map.
+    var subflightInputParameters = new FlightMap();
+    subflightInputParameters.put(JobMapKeys.AUTH_USER_INFO.getKeyName(), userRequest);
+    subflightInputParameters.put(
+        ControlledResourceKeys.DESTINATION_WORKSPACE_ID, destinationWorkspaceId);
+    subflightInputParameters.put(
+        ControlledResourceKeys.CLONING_INSTRUCTIONS, sourceResource.getCloningInstructions());
+    subflightInputParameters.put(ResourceKeys.RESOURCE, sourceResource);
+    subflightInputParameters.put(
+        JobMapKeys.DESCRIPTION.getKeyName(),
+        String.format(
+            "Clone Azure Storage Container %s", sourceResource.getResourceId().toString()));
+    subflightInputParameters.put(
+        ControlledResourceKeys.DESTINATION_CONTAINER_NAME,
+        String.format(
+            "clone-%s-%s", destinationWorkspaceId, sourceResource.getStorageContainerName()));
+    subflightInputParameters.put(
+        ControlledResourceKeys.DESTINATION_RESOURCE_ID, destinationResourceId);
+    subflightInputParameters.put(ControlledResourceKeys.DESTINATION_FOLDER_ID, destinationFolderId);
+    // Do not do the policy merge on the sub-object clone
+    subflightInputParameters.put(WorkspaceFlightMapKeys.MERGE_POLICIES, false);
+
+    // launch the flight
+    try {
+      context
+          .getStairway()
+          .submit(
+              subflightId,
+              CloneControlledAzureStorageContainerResourceFlight.class,
+              subflightInputParameters);
+    } catch (DuplicateFlightIdException unused) {
+      return StepResult.getStepResultSuccess();
+    } catch (DatabaseOperationException | StairwayExecutionException e) {
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, e);
+    }
+    return StepResult.getStepResultSuccess();
+  }
+
+  // Nothing to undo
+  @Override
+  public StepResult undoStep(FlightContext context) throws InterruptedException {
+    return StepResult.getStepResultSuccess();
```

Comment:
This step is launching the sub-flight which should be able to handle its own rollbacks in the event of a failure. An undo of this step would mean cancelling the subflight it launched, or I suppose it could try to launch a separate `DeleteContainerFlight`. I opted to follow what we do with GCP buckets and BQ datasets, but I don't know if that could leave us vulnerable to orphaned storage container resources.

Reply:
I don't think you get orphans. Even if the clone flight fails later for some reason, the parts that got done stay done. Clone is not guaranteed to be atomic.